# Lab 2 - IA340

## View the Data

In [1]:
import warnings
warnings.simplefilter('ignore')

import pandas as pd

df = pd.read_csv('s3://ia340-2025-fall-960915131/Electric_Vehicle_Population_Data.csv') # load the data from s3
df.head() #show the top 5 rows

VIN (1-10)     County          City State  Postal Code  Model Year    Make  \
0  5YJ3E1EB5K     Yakima        Yakima    WA      98901.0        2019   TESLA   
1  1C4RJXU67R     Kitsap  Port Orchard    WA      98367.0        2024    JEEP   
2  KNDCD3LD0N  Snohomish      Lynnwood    WA      98036.0        2022     KIA   
3  5UXKT0C37H       King        Auburn    WA      98001.0        2017     BMW   
4  1N4AZ0CP1D     Skagit  Mount Vernon    WA      98273.0        2013  NISSAN   

      Model                   Electric Vehicle Type  \
0   MODEL 3          Battery Electric Vehicle (BEV)   
1  WRANGLER  Plug-in Hybrid Electric Vehicle (PHEV)   
2      NIRO  Plug-in Hybrid Electric Vehicle (PHEV)   
3        X5  Plug-in Hybrid Electric Vehicle (PHEV)   
4      LEAF          Battery Electric Vehicle (BEV)   

  Clean Alternative Fuel Vehicle (CAFV) Eligibility  Electric Range  \
0           Clean Alternative Fuel Vehicle Eligible           220.0   
1             Not eligible due to low battery range            21.0   
2             Not eligible due to low battery range            26.0   
3             Not eligible due to low battery range            14.0   
4           Clean Alternative Fuel Vehicle Eligible            75.0   

   Base MSRP  Legislative District  DOL Vehicle ID  \
0        0.0                  15.0       347724772   
1        0.0                  35.0       272165288   
2        0.0                  32.0       203182584   
3        0.0                  30.0       349010287   
4        0.0                  40.0       131684150   

              Vehicle Location                               Electric Utility  \
0  POINT (-120.50729 46.60464)                                     PACIFICORP   
1  POINT (-122.68471 47.50524)                         PUGET SOUND ENERGY INC   
2  POINT (-122.29245 47.82557)                         PUGET SOUND ENERGY INC   
3   POINT (-122.23035 47.3074)  PUGET SOUND ENERGY INC||CITY OF TACOMA - (WA)   
4  POINT (-122.33891 48.41644)                         PUGET SOUND ENERGY INC   

   2020 Census Tract  
0       5.307700e+10  
1       5.303509e+10  
2       5.306105e+10  
3       5.303303e+10  
4       5.305795e+10

In [32]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 257027 entries, 0 to 257634
Data columns (total 17 columns):
 #   Column                                             Non-Null Count   Dtype  
---  ------                                             --------------   -----  
 0   VIN (1-10)                                         257027 non-null  object 
 1   County                                             257027 non-null  object 
 2   City                                               257027 non-null  object 
 3   State                                              257027 non-null  object 
 4   Postal Code                                        257027 non-null  float64
 5   Model Year                                         257027 non-null  int64  
 6   Make                                               257027 non-null  object 
 7   Model                                              257027 non-null  object 
 8   EV Type                                            257027 non-null  object 
 9 

In [33]:
df.describe()

Postal Code     Model Year  Electric Range      Base MSRP  \
count  257027.000000  257027.000000   257027.000000  257027.000000   
mean    98272.187292    2021.714139       43.115980     704.497271   
std       312.932246       3.024491       81.619388    6995.191558   
min     98001.000000    2000.000000        0.000000       0.000000   
25%     98052.000000    2020.000000        0.000000       0.000000   
50%     98133.000000    2023.000000        0.000000       0.000000   
75%     98382.000000    2024.000000       35.000000       0.000000   
max     99403.000000    2026.000000      337.000000  845000.000000   

       Legislative District  DOL Vehicle ID  2020 Census Tract  
count         257027.000000    2.570270e+05       2.570270e+05  
mean              28.897454    2.404200e+08       5.304013e+10  
std               14.882224    6.603734e+07       1.654160e+07  
min                1.000000    4.385000e+03       5.300195e+10  
25%               17.000000    2.134556e+08       5.303301e+10  
50%               32.000000    2.586898e+08       5.303303e+10  
75%               42.000000    2.738358e+08       5.305307e+10  
max               49.000000    4.792548e+08       5.307794e+10

In [35]:
df.shape

(257027, 17)

In [3]:
df.shape
df['City']

0               Yakima
1         Port Orchard
2             Lynnwood
3               Auburn
4         Mount Vernon
              ...     
257630        Mukilteo
257631        Kirkland
257632       Vancouver
257633      Bellingham
257634         Seattle
Name: City, Length: 257635, dtype: object

### Markdown for View Data: There are 257,027 rows and 17 columns. 5 columns are floats and 2 are integer so there are 7 total numeric columns. The other 10 are text. First impression is that there are various descriptive columns about each car.

## Explore and Clean Data

In [36]:
df = df.rename(columns={"Electric Vehicle Type": "EV Type"}) #rename one column, renamed "Electric Vehicle Type" to "EV Type"
df.head()

VIN (1-10)     County          City State  Postal Code  Model Year    Make  \
0  5YJ3E1EB5K     Yakima        Yakima    WA      98901.0        2019   TESLA   
1  1C4RJXU67R     Kitsap  Port Orchard    WA      98367.0        2024    JEEP   
2  KNDCD3LD0N  Snohomish      Lynnwood    WA      98036.0        2022     KIA   
3  5UXKT0C37H       King        Auburn    WA      98001.0        2017     BMW   
4  1N4AZ0CP1D     Skagit  Mount Vernon    WA      98273.0        2013  NISSAN   

      Model                                 EV Type  \
0   MODEL 3          Battery Electric Vehicle (BEV)   
1  WRANGLER  Plug-in Hybrid Electric Vehicle (PHEV)   
2      NIRO  Plug-in Hybrid Electric Vehicle (PHEV)   
3        X5  Plug-in Hybrid Electric Vehicle (PHEV)   
4      LEAF          Battery Electric Vehicle (BEV)   

  Clean Alternative Fuel Vehicle (CAFV) Eligibility  Electric Range  \
0           Clean Alternative Fuel Vehicle Eligible           220.0   
1             Not eligible due to low battery range            21.0   
2             Not eligible due to low battery range            26.0   
3             Not eligible due to low battery range            14.0   
4           Clean Alternative Fuel Vehicle Eligible            75.0   

   Base MSRP  Legislative District  DOL Vehicle ID  \
0        0.0                  15.0       347724772   
1        0.0                  35.0       272165288   
2        0.0                  32.0       203182584   
3        0.0                  30.0       349010287   
4        0.0                  40.0       131684150   

              Vehicle Location                               Electric Utility  \
0  POINT (-120.50729 46.60464)                                     PACIFICORP   
1  POINT (-122.68471 47.50524)                         PUGET SOUND ENERGY INC   
2  POINT (-122.29245 47.82557)                         PUGET SOUND ENERGY INC   
3   POINT (-122.23035 47.3074)  PUGET SOUND ENERGY INC||CITY OF TACOMA - (WA)   
4  POINT (-122.33891 48.41644)                         PUGET SOUND ENERGY INC   

   2020 Census Tract  
0       5.307700e+10  
1       5.303509e+10  
2       5.306105e+10  
3       5.303303e+10  
4       5.305795e+10

In [37]:
df= df.dropna() #drop nulls
df.isnull().sum() #check for null values

VIN (1-10)                                           0
County                                               0
City                                                 0
State                                                0
Postal Code                                          0
Model Year                                           0
Make                                                 0
Model                                                0
EV Type                                              0
Clean Alternative Fuel Vehicle (CAFV) Eligibility    0
Electric Range                                       0
Base MSRP                                            0
Legislative District                                 0
DOL Vehicle ID                                       0
Vehicle Location                                     0
Electric Utility                                     0
2020 Census Tract                                    0
dtype: int64

### Markdown for Clean the Data: I renamed the column "Electric Vehicle Type" to "EV Type". I dropped all nulls before checking how many there originally were, but I checked after dropping them to make sure there were 0. 

## Manipulate the Data

In [29]:
df['City'] #shows just one column

0               Yakima
1         Port Orchard
2             Lynnwood
3               Auburn
4         Mount Vernon
              ...     
257630        Mukilteo
257631        Kirkland
257632       Vancouver
257633      Bellingham
257634         Seattle
Name: City, Length: 257027, dtype: object

In [30]:
df.sort_values(by='Electric Range') #sort/order data by column values

VIN (1-10)     County       City State  Postal Code  Model Year  \
117879  5YJ3E1EA5S       King    Redmond    WA      98052.0        2025   
117896  7SAYGAEE1S  Snohomish   Lynnwood    WA      98037.0        2025   
117898  7SAYGDEEXP       King    Redmond    WA      98052.0        2023   
117885  5YJ3E1EB3S     Yakima     Yakima    WA      98902.0        2025   
117886  7PDSGABA2P       King  Sammamish    WA      98075.0        2023   
...            ...        ...        ...   ...          ...         ...   
19913   5YJSA1E41L    Spokane    Spokane    WA      99223.0        2020   
171519  5YJSA1E43L   Thurston    Olympia    WA      98516.0        2020   
69872   5YJSA1E43L      Grant     Quincy    WA      98848.0        2020   
184661  5YJSA1E45L    Cowlitz   Longview    WA      98632.0        2020   
140589  5YJSA1E45L      Clark      Camas    WA      98607.0        2020   

          Make    Model                         EV Type  \
117879   TESLA  MODEL 3  Battery Electric Vehicle (BEV)   
117896   TESLA  MODEL Y  Battery Electric Vehicle (BEV)   
117898   TESLA  MODEL Y  Battery Electric Vehicle (BEV)   
117885   TESLA  MODEL 3  Battery Electric Vehicle (BEV)   
117886  RIVIAN      R1S  Battery Electric Vehicle (BEV)   
...        ...      ...                             ...   
19913    TESLA  MODEL S  Battery Electric Vehicle (BEV)   
171519   TESLA  MODEL S  Battery Electric Vehicle (BEV)   
69872    TESLA  MODEL S  Battery Electric Vehicle (BEV)   
184661   TESLA  MODEL S  Battery Electric Vehicle (BEV)   
140589   TESLA  MODEL S  Battery Electric Vehicle (BEV)   

        Clean Alternative Fuel Vehicle (CAFV) Eligibility  Electric Range  \
117879  Eligibility unknown as battery range has not b...             0.0   
117896  Eligibility unknown as battery range has not b...             0.0   
117898  Eligibility unknown as battery range has not b...             0.0   
117885  Eligibility unknown as battery range has not b...             0.0   
117886  Eligibility unknown as battery range has not b...             0.0   
...                                                   ...             ...   
19913             Clean Alternative Fuel Vehicle Eligible           337.0   
171519            Clean Alternative Fuel Vehicle Eligible           337.0   
69872             Clean Alternative Fuel Vehicle Eligible           337.0   
184661            Clean Alternative Fuel Vehicle Eligible           337.0   
140589            Clean Alternative Fuel Vehicle Eligible           337.0   

        Base MSRP  Legislative District  DOL Vehicle ID  \
117879        0.0                  48.0       273975019   
117896        0.0                  32.0       274099101   
117898        0.0                  48.0       254092429   
117885        0.0                  15.0       281764500   
117886        0.0                  41.0       276276955   
...           ...                   ...             ...   
19913         0.0                   6.0       132182620   
171519        0.0                  22.0       102849792   
69872         0.0                  13.0       278275833   
184661        0.0                  19.0       132280116   
140589        0.0                  18.0       262297661   

                   Vehicle Location  \
117879  POINT (-122.13158 47.67858)   
117896  POINT (-122.27734 47.83785)   
117898  POINT (-122.13158 47.67858)   
117885  POINT (-120.52041 46.59751)   
117886  POINT (-122.03539 47.61344)   
...                             ...   
19913   POINT (-117.36043 47.63396)   
171519  POINT (-122.75103 47.08149)   
69872   POINT (-119.85338 47.23748)   
184661  POINT (-122.95058 46.14681)   
140589  POINT (-122.40199 45.58694)   

                                         Electric Utility  2020 Census Tract  
117879      PUGET SOUND ENERGY INC||CITY OF TACOMA - (WA)       5.303302e+10  
117896                             PUGET SOUND ENERGY INC       5.306105e+10  
117898      PUGET SOUND ENERGY INC||CITY OF TACOMA - (WA) 

In [10]:
df[0:5] #slice by rows (shows first 5 rows)

VIN (1-10)     County          City State  Postal Code  Model Year    Make  \
0  5YJ3E1EB5K     Yakima        Yakima    WA      98901.0        2019   TESLA   
1  1C4RJXU67R     Kitsap  Port Orchard    WA      98367.0        2024    JEEP   
2  KNDCD3LD0N  Snohomish      Lynnwood    WA      98036.0        2022     KIA   
3  5UXKT0C37H       King        Auburn    WA      98001.0        2017     BMW   
4  1N4AZ0CP1D     Skagit  Mount Vernon    WA      98273.0        2013  NISSAN   

      Model                   Electric Vehicle Type  \
0   MODEL 3          Battery Electric Vehicle (BEV)   
1  WRANGLER  Plug-in Hybrid Electric Vehicle (PHEV)   
2      NIRO  Plug-in Hybrid Electric Vehicle (PHEV)   
3        X5  Plug-in Hybrid Electric Vehicle (PHEV)   
4      LEAF          Battery Electric Vehicle (BEV)   

  Clean Alternative Fuel Vehicle (CAFV) Eligibility  Electric Range  \
0           Clean Alternative Fuel Vehicle Eligible           220.0   
1             Not eligible due to low battery range            21.0   
2             Not eligible due to low battery range            26.0   
3             Not eligible due to low battery range            14.0   
4           Clean Alternative Fuel Vehicle Eligible            75.0   

   Base MSRP  Legislative District  DOL Vehicle ID  \
0        0.0                  15.0       347724772   
1        0.0                  35.0       272165288   
2        0.0                  32.0       203182584   
3        0.0                  30.0       349010287   
4        0.0                  40.0       131684150   

              Vehicle Location                               Electric Utility  \
0  POINT (-120.50729 46.60464)                                     PACIFICORP   
1  POINT (-122.68471 47.50524)                         PUGET SOUND ENERGY INC   
2  POINT (-122.29245 47.82557)                         PUGET SOUND ENERGY INC   
3   POINT (-122.23035 47.3074)  PUGET SOUND ENERGY INC||CITY OF TACOMA - (WA)   
4  POINT (-122.33891 48.41644)                         PUGET SOUND ENERGY INC   

   2020 Census Tract  
0       5.307700e+10  
1       5.303509e+10  
2       5.306105e+10  
3       5.303303e+10  
4       5.305795e+10

In [31]:
df[df['Electric Range']>5] #shows only rows where electric range > 5 (filtering with a condition)

VIN (1-10)     County          City State  Postal Code  Model Year  \
0       5YJ3E1EB5K     Yakima        Yakima    WA      98901.0        2019   
1       1C4RJXU67R     Kitsap  Port Orchard    WA      98367.0        2024   
2       KNDCD3LD0N  Snohomish      Lynnwood    WA      98036.0        2022   
3       5UXKT0C37H       King        Auburn    WA      98001.0        2017   
4       1N4AZ0CP1D     Skagit  Mount Vernon    WA      98273.0        2013   
...            ...        ...           ...   ...          ...         ...   
257626  5YJ3E1EB5L    Spokane       Spokane    WA      99212.0        2020   
257628  WP1AE2A27F  Snohomish      Mukilteo    WA      98275.0        2015   
257630  2C4RC1L73L  Snohomish      Mukilteo    WA      98275.0        2020   
257632  WBAJB1C53K      Clark     Vancouver    WA      98683.0        2019   
257633  JTDKARFP9K    Whatcom    Bellingham    WA      98229.0        2019   

            Make               Model                                 EV Type  \
0          TESLA             MODEL 3          Battery Electric Vehicle (BEV)   
1           JEEP            WRANGLER  Plug-in Hybrid Electric Vehicle (PHEV)   
2            KIA                NIRO  Plug-in Hybrid Electric Vehicle (PHEV)   
3            BMW                  X5  Plug-in Hybrid Electric Vehicle (PHEV)   
4         NISSAN                LEAF          Battery Electric Vehicle (BEV)   
...          ...                 ...                                     ...   
257626     TESLA             MODEL 3          Battery Electric Vehicle (BEV)   
257628   PORSCHE             CAYENNE  Plug-in Hybrid Electric Vehicle (PHEV)   
257630  CHRYSLER            PACIFICA  Plug-in Hybrid Electric Vehicle (PHEV)   
257632       BMW                530E  Plug-in Hybrid Electric Vehicle (PHEV)   
257633    TOYOTA  PRIUS PRIME (PHEV)  Plug-in Hybrid Electric Vehicle (PHEV)   

       Clean Alternative Fuel Vehicle (CAFV) Eligibility  Electric Range  \
0                Clean Alternative Fuel Vehicle Eligible           220.0   
1                  Not eligible due to low battery range            21.0   
2                  Not eligible due to low battery range            26.0   
3                  Not eligible due to low battery range            14.0   
4                Clean Alternative Fuel Vehicle Eligible            75.0   
...                                                  ...             ...   
257626           Clean Alternative Fuel Vehicle Eligible           322.0   
257628             Not eligible due to low battery range            14.0   
257630           Clean Alternative Fuel Vehicle Eligible            32.0   
257632             Not eligible due to low battery range            15.0   
257633             Not eligible due to low battery range            25.0   

        Base MSRP  Legislative District  DOL Vehicle ID  \
0             0.0                  15.0       347724772   
1             0.0                  35.0       272165288   
2             0.0                  32.0       203182584   
3             0.0                  30.0       349010287   
4             0.0                  40.0       131684150   
...           ...                   ...             ...   
257626        0.0                   4.0         4783222   
257628        0.0                  21.0       143946596   
257630        0.0                  21.0       161576068   
257632    55700.0                  18.0       278137940   
257633        0.0                  40.0       177812232   

                   Vehicle Location  \
0       POINT (-120.50729 46.60464)   
1       POINT (-122.68471 47.50524)   
2       POINT (-122.29245 47.82557)   
3        POINT (-122.23035 47.3074)   
4       POINT (-122.33891 48.41644)   
...                             ...   
257626  POINT (-117.28805 47.68043)   
257628  POINT (-122.29196 47.89908)   
257630  POINT (-122.29196 47.89908)   
257632  POINT (-122.49212 45.60365)   
257633   POINT (-122.45486 48.7449)   

                   

### I showed the column "City". I sorted values by the column "Electric Range". I sliced the first 5 rows. I filtered rows by the coniditon of "Elecric Range" > 5. 